mlir/test/Dialect/Arith/canonicalize.mlir

```diff
@@ -1051,6 +1193,17 @@ func.func @addiMuliToSubiLhsI32(%arg0: i32, %arg1: i32) -> i32 {
   return %add : i32
 }
 
+// CHECK-LABEL: @addiMuliToSubiLhsI32Ovf
+//  CHECK-SAME:   (%[[ARG0:.+]]: i32, %[[ARG1:.+]]: i32)
+//       CHECK:   %[[SUB:.+]] = arith.subi %[[ARG0]], %[[ARG1]] overflow<nsw, nuw> : i32
+//       CHECK:   return %[[SUB]]
+func.func @addiMuliToSubiLhsI32Ovf(%arg0: i32, %arg1: i32) -> i32 {
```

Comment:
When running this through alive I get a mismatch:
https://alive2.llvm.org/ce/z/8NhXoT
I did enter things manually so apologies if I made a mistake. It seems like the multiplication of 0 with -1 does not result in poison (despite the nuw flag) while sub 0, 1 does result in a unsigned overflow. The result is a bit of a surprise for me but maybe others can shed some light on it?

Could make to run this tests systematically through alive? I did it manually but it could be fairly quick to lower the test input and output to llvm ir and the copy past these overflow flag tests individually into alive. I have more trust in a tool than in my reviewing skills here. What are your takes on this @kuhar @Hardcode84?

Reply:
+1, some script which can run entire arith canonicalization suite through alive2 will be extremely useful. I can think about it and switch this PR to draft fro now.

Reply:
So, I wrote this abomination https://gist.github.com/Hardcode84/13a8c5a45aa0e3849e1ab5d30a56749e
* `-split-input-file ` doesn't work properly and I had to manually remove `// -----` markers
* Had to remove `tensor` mentions from the file as it cannot be translated
* Had to remove i0 tests as they hard-crash `-mlir-to-llvmir`
* Had to manually cleanup duplicated `source_filename` and metadata from result.

And all of this was in vain as alive2 timeout on such big input https://alive2.llvm.org/ce/z/bjvhLR

Reply:
I guess they website has limited compute time allocated :(. 

I don't think the script is in vain though. I would start by running it on the new tests you added. Worst case test case by test case to avoid the time out.

Reply:
Here is better script https://gist.github.com/Hardcode84/bbf08e0923d4b9004aa831d0a1d3ac4f

output https://alive2.llvm.org/ce/z/KhQs4J

Reply:
Nice!

Looks like the patterns that convert add <-> sub seem to be tricky, probably because addition uses an unsigned overflow to implement subtraction? So maybe dropping the nuw flag helps.

Reply:
I'll remove patterns that failed from this PR and will create separate PRs for them and for the script.